Implemented PWA

alistair3149 opened the pull request:
Implemented requirements needed for Progressive Web App.
ServiceWorker is working but needed more work on caching API JSON files so that the whole site is usable in offline mode. But for now it is good as a start.

alistair3149 commented:
Yes the service worker cached the pages and their assets that you visited before. So let's say that I only visited the Carrack page and I went offline, I'll be able to visit the Carrack page again with the data there, and other ship pages as well as they use the same assets, but no data will be there as it's never retrieved.

As for the second question, the API age is separated so it can be set lowered to 5 mins and the rest for longer.

alistair3149 commented:
I updated the SW to check to fetch from network first and use the cache version after a timeout of 15 seconds (it can be lower to increase responsiveness, but depends on the fetch time of the API usually). I also switched the max age to 3 days instead, the old max age was a copy and paste from the SO and I didn't think that though >:(

richardthombs commented:
Can you separate out the non-service-worker related commits into a different pull request? If I merge this, I have no choice but to also suck in all the UI changes and I'm not ready to do that yet :)

alistair3149 commented:
There is no UI change in this PR, all the UI changes are in #15 .

richardthombs commented on website/ngsw-config.json:
```diff
@@ -0,0 +1,44 @@
+      ],
+      "cacheConfig": {
+        "strategy": "freshness",
+        "maxSize": 100,
```
I think 100 is too small. If you load the Carrack up while the dev tools are open, there are 74 requests. Something small like the Argo has 26. So I think maybe 20 ships * 50 requests per ship, we should have 1,000 as a minimum.

alistair3149 replied:
Increased to 1,000 in the latest commit

richardthombs commented:
I don't know if you are seeing the same data, but this is what I see in the PR:

![image](https://user-images.githubusercontent.com/1511599/78059315-88201200-7381-11ea-8b8c-bc773079c95d.png)

alistair3149 commented:
All the commit are already merged into the master branch with the other PR, the duplicated one is because I rebased the branch. For the second commit, I changed the theme color to deeppink from the angular blue as that seems to be the main color on the site, it does not change the UI other than the browser and status bar color when the site is loaded on mobile.

alistair3149 commented:
Closed for now, will reopen after cleaning up the commit

alistair3149 commented:
Cleaned up commit and re-opened PR at #16
